test/blackbox-tests/test-cases/merlin/granularity.t

```diff
@@ -0,0 +1,178 @@
+  $ ocamlc_where="$(ocamlc -where)"
+  $ export BUILD_PATH_PREFIX_MAP="/OCAMLC_WHERE=$ocamlc_where:$BUILD_PATH_PREFIX_MAP"
+
+# A utility to query merlin configuration for a file:
+  $ cat >merlin_conf.sh <<EOF
+  > #!/bin/sh
+  > FILE=\$1
+  > printf "(4:File%d:%s)" \${#FILE} \$FILE | dune ocaml-merlin \
+  >   | sed -E "s/[[:digit:]]+:/\?:/g" | sed -E "s/\)(\(+)/\n\1/g"
+  > EOF
+
+  $ chmod a+x merlin_conf.sh
+
+# Project sources
+
+  $ cat >dune-project <<EOF
+  > (lang dune 3.11)
+  > (using melange 0.1)
+  > 
+  > (dialect
+  >  (name mlx)
+  >  (implementation
+  >   (extension mlx)
+  >   (preprocess (run cat %{input-file}))
+  >   (merlin_reader mlx)))
+  > EOF
+
+  $ cat >pp.sh <<EOF
+  > #!/bin/sh
+  > sed 's/%INT%/42/g' \$1
+  > EOF
+
+  $ chmod a+x pp.sh
+
+  $ cat >dune <<EOF
+  > (executable
+  >  (name test)
+  >  (flags :standard -no-strict-formats)
+  >  (preprocess
+  >   (per_module
+  >    ((action (run ./pp.sh %{input-file})) pped))))
+  > 
+  > (rule
+  >  (action  (copy cppomod.cppo.ml cppomod.ml)))
+  > 
+  > (rule
+  >  (action  (copy wrongext.cppo.cml wrongext.ml)))
+  > EOF
+
+  $ cat >test.ml <<EOF
+  > print_endline Pped.x_42;;
+  > print_endline Mel.x;;
+  > print_endline Cppomod.x;;
+  > print_endline Wrongext.x;;
+  > EOF
+
+Both Pped's ml and mli files will be preprocessed
+  $ cat >pped.mli <<EOF
+  > val x_%INT% : string
+  > EOF
+
+  $ cat >pped.ml <<EOF
+  > let x_42 = "%INT%"
+  > EOF
+
+Melange module signature
+  $ cat >mel.mli <<EOF
+  > val x : string
+  > EOF
+
+Melange module implementation in Melange syntax
+  $ cat >mel.mlx <<EOF
+  > let x = "43"
+  > EOF
+
+A pped file with unconventionnal filename
+  $ cat >cppomod.cppo.ml <<EOF
+  > let x = "44"
+  > EOF
+
+  $ cat >wrongext.cppo.cml <<EOF
+  > let x = "45"
+  > EOF
+
+  $ dune build @check
+  $ dune exec ./test.exe
+  42
+  43
+  44
+  45
+
+# We now query Merlin configuration for the various source files:
+
+Some configuration fields are common to all the modules of a same stanza. This
+is the case for the stdlib, build and sources directories, flags and suffixes.
+
+Some configuration can be specific to a module like preprocessing.
+
+Dialects are specified by extensions so are specific to a file. This means a
+different reader might be used for the signature and the implementation.
+
+Note that suffixes provided by dialect should always be told to Merlin for
+locate to work correctly. 
+
+## Preprocessing:
+
+Is it expected that the suffix for implementation and interface is the same ?
+  $ ./merlin_conf.sh pped.ml | tee pped.out
+  ((?:STDLIB?:/OCAMLC_WHERE
+  (?:EXCLUDE_QUERY_DIR
+  (?:B?:$TESTCASE_ROOT/_build/default/.test.eobjs/byte
+  (?:S?:$TESTCASE_ROOT
+  (?:FLG(?:-open?:Dune__exe)
+  (?:FLG(?:-pp?:$TESTCASE_ROOT/_build/default/pp.sh)
+  (?:FLG(?:-w?:@1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-?:-strict-sequence?:-strict-formats?:-short-paths?:-keep-locs?:-no-strict-formats?:-g)
+  (?:SUFFIX?:.mlx .mlx))
```

Comment:
what does `(?:SUFFIX?:.mlx .mlx))` mean here? Is `pped.ml` getting read by merlin as if it belonged to the mlx dialect?

Reply:
merlin uses `SUFFIX` config to determine if this is an implementation or an interface document

now that `.mlx .mlx` is not completely correct as mlx dialect doesn't really have an interface counterpart to `.mlx` and it is supposed to use `.mli` instead

@voodoos suggested [here](https://github.com/ocaml/dune/pull/8567#pullrequestreview-1624199713) to use `.mlx .mli` which I want to do but in a separate PR (the code is [here](https://github.com/andreypopp/dune/blob/a5fc103dda82cffff1985dcfff8955ff035b12be/src/dune_rules/merlin/merlin.ml#L144-L150)) as it seems another kind of change.